### PR TITLE
raft topology: reject replace if the node being replaced is not dead

### LINKED ISSUE #15863: raft topology: join: check that the node being replaced is dead
Sub-task of https://github.com/scylladb/scylladb/issues/15555, followup to https://github.com/scylladb/scylladb/pull/15196#discussion_r1336912430.

Relevant code from the topology coordinator:

```c++
    std::variant<join_node_response_params::accepted, join_node_response_params::rejected>
    validate_joining_node(const node_to_work_on& node) {
        if (node.rs->state == node_state::replacing) {
            auto replaced_id = std::get<replace_param>(node.req_param.value()).replaced_id;
            if (!_topo_sm._topology.normal_nodes.contains(replaced_id)) {
                return join_node_response_params::rejected {
                    .reason = ::format("Cannot replace node {} because it is not in the 'normal' state", replaced_id),
                };
            }
        }
    }
```

We should check here that the node we are trying to replace is dead.

### PR DESCRIPTION
**Pull request by patjed41**
The replace operation is defined to succeed only if the node being
replaced is dead. We should reject this operation when the failure
detector considers the node being replaced alive.

Apart from adding this change, this PR adds a test case -
`test_replacing_alive_node_fails` - that verifies it. A few testing
framework adjustments were necessary to implement this test and
to avoid flakiness in other tests that use the replace operation after
the change. From now, we need to ensure that all nodes see the
node being replaced as dead before starting the replace. Otherwise,
the check added in this PR could reject the replace.

Additionally, this PR changes the replace procedure in a way that
if the replacing node reuses the IP of the node being replaced, other
nodes can see it as alive only after the topology coordinator accepts
its join request. The replacing node may become alive before the
topology coordinator checks if the node being replaced is dead. If
that happens and the replacing node reuses the IP of the node being
replaced, the topology coordinator cannot know which of these two
nodes is alive and whether it should reject the join request.

Fixes #15863

### CODE REVIEW

**Comment by avikivity** [comment]
missing: test

**Comment by kostja** [comment]
Can we have a test please?

**Review by kostja — CHANGES_REQUESTED** [review]
Can we have a test please?

**Comment by patjed41** [comment]
I'm currently working on adding a test.

A natural solution is to add a test case to `test_replace.py`. However, there are already many test cases, and all of them unnecessarily start a new cluster, so I tried to merge them into one test case. Then, I discovered #15967.

I also discovered a small bug in `topology_coordinator::validate_joining_node`. The condition in `if (node.rs->state == node_state::replacing)` is incorrect. `node.rs->state` always evaluates to `node_state::none` for both join and replace. The fix is very simple -- just changing the condition. I'll add a commit with it.

Finally, the testing framework doesn't allow replacing a running node. I need to make a few changes, and after that, I'll be able to submit a test.

**Comment by patjed41** [comment]
v2:
- left the two commits from v1 unchanged (commits 6, 7)
- fixed a bug in `topology_coordinator::validate_joining_node` (commit 1)
- delayed advertising the replacing node if it reuses IP of the node being replaced (commit 2); without this change checking if the node being replaced is dead was bugged
- added a test (commit 8)
- made a few adjustments in the testing framework necessary to write a test and to avoid flakiness in other tests that use the replace operation (commits 3, 4, 5)
- updated PR's description
- rebased on the updated master

File: test/pylib/manager_client.py
Comment (kbr-scylla) [inline test/pylib/manager_client.py:177]: Re commit message
```
The first one is that if we want to add a server while expecting
an error, we have to do it in two steps:
- call server_add with the start parameter set to False,
- call server_stop with the expected_error parameter.
```

I think you wanted `server_start` instead of `server_stop`.

File: test/topology/test_replace.py
Comment (kbr-scylla) [inline test/topology/test_replace.py:54]: I'd suggest to enable this test only in a single mode like `dev` to save some resources, we don't really exercise some critical paths here so no need to run it in `debug`.

**Comment by kbr-scylla** [comment]
@gleb-cloudius @piodul please review

File: test/pylib/manager_client.py
Comment (piodul) [inline test/pylib/manager_client.py:199]: I don't like that we are making the wait conditional on whether we expect an error or not. It's not obvious that it's always what we want - what if we want to write a different test than `test_replacing_alive_node_fails` later which expects replace to fail, but not due to the node still being seen as alive by others?

Instead, I think we should extend `ReplaceConfig` to allow choosing whether we want to wait or not; alternatively, add a parameter to `add_server` which allows to control it.

File: service/storage_service.cc
Comment (gleb-cloudius) [reply service/storage_service.cc:2978]: Because we do not need to put tokens into the `gossiper state if _raft_topology_change_enabled == true` and before the patch we did not because `replacing_a_node_with_same_ip` and `replacing_a_node_with_diff_ip` were false.

File: test/topology/test_replace.py
Comment (patjed41) [reply test/topology/test_replace.py:54]: Right, this test is also not the fastest. I'll move it to a separate file and make it `dev` only.

File: test/pylib/manager_client.py
Comment (patjed41) [reply test/pylib/manager_client.py:199]: Sure, I'll change it. I think I'll take the second approach. I don't think this waiting is logically a part of the replace config. Also, `server_start` and `server_restart` already have a similar `wait_others` parameter.

File: service/storage_service.cc
Comment (patjed41) [reply service/storage_service.cc:2978]: I think this change should be explained in the commit message. I'll add it.

Comment (patjed41) [reply service/storage_service.cc:2978]: Done.

File: test/pylib/manager_client.py
Comment (patjed41) [reply test/pylib/manager_client.py]: Done.

**Comment by patjed41** [comment]
v2:
- extended and fixed commit messages
- made waiting until the node being replaced is dead in `ManagerClient` optional
- moved `test_replacing_alive_node_fails` to a separate file and made it `dev` only

**Comment by patjed41** [comment]
rebased on a new master as my next PR will depend on this one

**Review by piodul — APPROVED** [review]

File: test/pylib/manager_client.py
Comment (kbr-scylla) [inline test/pylib/manager_client.py:183]: This looks like it belongs inside `ReplaceConfig` struct

File: test/topology/test_cluster_features.py
Comment (kbr-scylla) [inline test/topology/test_cluster_features.py:148]: I don't like this.

We should be testing cluster behavior under abrupt failures (e.g. `kill -9`), we can't rely on graceful shutdowns to happen each time.

Each time I review this PR I have more and more of this feeling that we're doing something wrong...

First of all, using gossiper liveness information is of course unreliable. What happens when node is marked down by gossiper, not because the node is down, but is behind a network partition? What will the topology coordinator do in this case? I guess that it will attempt to do the replace operation and ban the node in the process. What if the node comes back in the middle of replace? Will it disrupt the cluster while replace is ongoing? Will it send/receive data path operations? Maybe the correct way to handle it is to ban it *at the beginning* of replace operation, not at the end as it is now (if I remember correctly)? @gleb-cloudius WDYT?

I think it's understood that we should be using this stuff only for improving user experience, not relying on it for safety (which is impossible).

So fine, we can:
- reject replace if node is considered alive
- and still have a mechanism that protects us (like early node ban) in case we replace an actually alive (or just temporarily dead) node.

Please check if we need to do that second thing, if it's not done already (do the experiment with having a replaced node come back to life in the middle / at the beginning of replace). If not, please create another follow-up issue.

Second, regarding the abrupt stop. We should have tests with abrupt stop. At least one. Of course waiting 20 seconds because of the user-experience gossiper-based protection is not an option. I see the following options:
- this 20 seconds delay is failure detector setting, which can be configured. We could move these replace tests into `topology_custom` and start clusters there with custom configuration, reducing the delay. This should go together with disabling the tests in debug mode, where the low delay combined with latencies may cause failure detector "flickering"
- OR disable the protection using a failure injection, but that probably also requires moving the tests to `topology_custom`, to enable the failure injection when cluster starts...
- do the above but only to some tests (like the replace tests), the non-replace-specific tests here can keep using `server_stop_gracefully`.

Note I don't want to increase the scope of this PR so that I would also leave to a follow-up issue, which is probably low priority, but still it's an issue. Please create one after the PR is merged.

Comment (gleb-cloudius) [reply test/topology/test_cluster_features.py:148]: > I don't like this.
> 
> We should be testing cluster behavior under abrupt failures (e.g. `kill -9`), we can't rely on graceful shutdowns to happen each time.
> 
> Each time I review this PR I have more and more of this feeling that we're doing something wrong...
> 
> First of all, using gossiper liveness information is of course unreliable. What happens when node is marked down by gossiper, not because the node is down, but is behind a network partition? What will the topology coordinator do in this case? I guess that it will attempt to do the replace operation and ban the node in the process. What if the node comes back in the middle of replace? Will it disrupt the cluster while replace is ongoing? Will it send/receive data path operations? Maybe the correct way to handle it is to ban it _at the beginning_ of replace operation, not at the end as it is now (if I remember correctly)? @gleb-cloudius WDYT?

Of course. This was always the final plan. We also need to ban all nodes in the ignored_node list. We just do not have a mechanism to do it since the one we have relies on the `left` state (IIRC).

Comment (kbr-scylla) [reply test/topology/test_cluster_features.py:148]: If something is in "the final plan", we should have an issue about it, not keep it in one or two people's heads. (It definitely skipped my awareness until now.) I don't think we have one. @patjed41 please do create an issue about this.

Comment (gleb-cloudius) [reply test/topology/test_cluster_features.py:148]: > If something is in "the final plan", we should have an issue about it, not keep it in one or two people's heads. (It definitely skipped my awareness until now.) I don't think we have one. @patjed41 please do create an issue about this.

At least ignored_nodes isolation is in the failure recovery doc:

```
def topology_change():
   isolate_nodes(ignored_nodes)
   version++
```

The plan was to reuse the same `isolate_nodes` that we were going to use during a failure while recovering from a previous one, but in the end that was postponed/canceled, so we did not write `isolate_nodes`.

Comment (patjed41) [reply test/topology/test_cluster_features.py:148]: Do we have the same plan for removenode? That is, banning the node being removed and the ignored nodes. If yes, I think we can create one issue for both replace and removenode.

If I remember correctly, we also want to reject removenode if the node being removed is alive. We already do this in `storage_service::raft_removenode`, but I think we should also (or only) do it on the topology coordinator's side since the gossip state might change. If I'm right, should I create an issue for this?

Comment (gleb-cloudius) [reply test/topology/test_cluster_features.py:148]: > Do we have the same plan for removenode?

Of course.

> That is, banning the node being removed and the ignored nodes. If yes, I think we can create one issue for both replace and removenode.
> 
> If I remember correctly, we also want to reject removenode if the node being removed is alive. We already do this in `storage_service::raft_removenode`, but I think we should also (or only) do it on the topology coordinator's side since the gossip state might change. If I'm right, should I create an issue for this?

Yes.

Comment (patjed41) [reply test/topology/test_cluster_features.py:148]: > Second, regarding the abrupt stop. We should have tests with abrupt stop. At least one. Of course waiting 20 seconds because of the user-experience gossiper-based protection is not an option. I see the following options:
> - this 20 seconds delay is failure detector setting, which can be configured. We could move these replace tests into `topology_custom` and start clusters there with custom configuration, reducing the delay. This should go together with disabling the tests in debug mode, where the low delay combined with latencies may cause failure detector "flickering"
> - OR disable the protection using a failure injection, but that probably also requires moving the tests to `topology_custom`, to enable the failure injection when cluster starts...
> - do the above but only to some tests (like the replace tests), the non-replace-specific tests here can keep using `server_stop_gracefully`.

I like the dot 1 + dot 3 approach the most. By "the replace tests" you probably mean `test_replace.py` and `test_raft_ignore_nodes.py`. That will leave us with only one `server_stop -> server_stop_gracefully` change but it's fine.

Comment (patjed41) [reply test/topology/test_cluster_features.py:148]: I opened #16108 and #16109

**Comment by patjed41** [comment]
v4:
- moved `wait_replaced_dead` to `ReplaceConfig`
- moved `test_replace.py` to `topology_custom`
- made all initial servers in `test_replace.py` and `test_raft_replace_ignore_nodes` use `failure_detector_timeout_in_ms = 2000`
- changed `server_stop_gracefully` calls back to `server_stop` in `test_replace.py` and `test_raft_replace_ignore_nodes`
- made `test_replace.py` skipped in the debug mode

**Review by kbr-scylla — APPROVED** [review]

File: test/topology/test_cluster_features.py
Comment (kbr-scylla) [reply test/topology/test_cluster_features.py:148]: > I opened https://github.com/scylladb/scylladb/issues/16108 and https://github.com/scylladb/scylladb/issues/16109

Great, thanks.

**Comment by kbr-scylla** [comment]
Waiting for next to stabilize before merged.

**Pull request merged**
